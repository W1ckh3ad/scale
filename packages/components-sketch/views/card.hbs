{{#* inline "icon"}}m12 1c6.1 0 11 4.9 11 11s-4.9 11-11 11-11-4.9-11-11 4.9-11 11-11zm0 1.5c-5.25 0-9.5 4.25-9.5 9.5s4.25 9.5 9.5 9.5 9.5-4.25 9.5-9.5-4.25-9.5-9.5-9.5zm-4.05 5.45c.3-.3.75-.3 1.05 0l3 3 3-3c.3-.3.75-.3 1.05 0s.3.75 0 1.05l-3 3 3 3c.3.3.3.75 0 1.05s-.75.3-1.05 0l-3-3-3 3c-.3.3-.75.3-1.05 0s-.3-.75 0-1.05l3-3-3-3c-.3-.3-.3-.75 0-1.05z{{/inline}}

<div data-sketch-artboard="Card" class="artboard" style="width: 1810px">
    <h1 class="title--artboard">Card</h1>
    
    <h2 class="title--artboard-section">Standard</h2>
    <div class="agent-states--grid">
        <div class="agent-states--item">
            <p class="agent-states--label">:enabled</p>
            <scale-card data-sketch-key="card-1">
                <div>Card Label</div>
            </scale-card>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:hover</p>
            <scale-card data-sketch-state="hover" data-sketch-key="card-2">
                <div>Card Label</div>
            </scale-card>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:focus</p>
            <scale-card data-sketch-state="focus" data-sketch-key="card-3">
                <div>Card Label</div>
            </scale-card>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:active/:selected</p>
            <scale-card data-sketch-state="active" data-sketch-key="card-4">
                <div>Card Label</div>
            </scale-card>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:disabled</p>
            <scale-card disabled="true" data-sketch-key="card-5">
                <div>Card Label</div>
            </scale-card>
        </div>
    </div>

    <h2 class="title--artboard-section">With Link</h2>
    <div class="agent-states--grid">
        <div class="agent-states--item">
            <p class="agent-states--label">:enabled</p>
            <scale-card to="https://example.com" target="_blank" rel="noopener noreferrer" data-sketch-key="card-6">
                <div>Card Label</div>
            </scale-card>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:hover</p>
            <scale-card to="https://example.com" target="_blank" rel="noopener noreferrer"  data-sketch-state="hover" data-sketch-key="card-7">
                <div>Card Label</div>
            </scale-card>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:focus</p>
            <scale-card to="https://example.com" target="_blank" rel="noopener noreferrer"  data-sketch-state="focus" data-sketch-key="card-8">
                <div>Card Label</div>
            </scale-card>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:active/:selected</p>
            <scale-card to="https://example.com" target="_blank" rel="noopener noreferrer"  data-sketch-state="active" data-sketch-key="card-9">
                <div>Card Label</div>
            </scale-card>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:disabled</p>
            <scale-card to="https://example.com" target="_blank" rel="noopener noreferrer"  disabled="true" data-sketch-key="card-10">
                <div>Card Label</div>
            </scale-card>
        </div>
    </div>

    <h2 class="title--artboard-section">With Image</h2>
    <div class="agent-states--grid">
        <div class="agent-states--item">
            <p class="agent-states--label">:enabled</p>
            <scale-card to="https://example.com" target="_blank" rel="noopener noreferrer" data-sketch-key="card-11">
                <div>
                    <div style="margin: -24px;">
                        <img src="/placeholder.png" alt="placeholder" style="margin-bottom: -4px; width: 100%;">
                        <div style="padding: 24px;">A card with a full width image</div>
                    </div>
                </div>
            </scale-card>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:hover</p>
            <scale-card to="https://example.com" target="_blank" rel="noopener noreferrer"  data-sketch-state="hover" data-sketch-key="card-12">
                <div>
                    <div style="margin: -24px;">
                        <img src="/placeholder.png" alt="placeholder" style="margin-bottom: -4px; width: 100%;">
                        <div style="padding: 24px;">A card with a full width image</div>
                    </div>
                </div>
            </scale-card>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:focus</p>
            <scale-card to="https://example.com" target="_blank" rel="noopener noreferrer"  data-sketch-state="focus" data-sketch-key="card-13">
                <div>
                    <div style="margin: -24px;">
                        <img src="/placeholder.png" alt="placeholder" style="margin-bottom: -4px; width: 100%;">
                        <div style="padding: 24px;">A card with a full width image</div>
                    </div>
                </div>
            </scale-card>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:active/:selected</p>
            <scale-card to="https://example.com" target="_blank" rel="noopener noreferrer"  data-sketch-state="active" data-sketch-key="card-14">
                <div>
                    <div style="margin: -24px;">
                        <img src="/placeholder.png" alt="placeholder" style="margin-bottom: -4px; width: 100%;">
                        <div style="padding: 24px;">A card with a full width image</div>
                    </div>
                </div>
            </scale-card>
        </div>
        <div class="agent-states--item">
            <p class="agent-states--label">:disabled</p>
            <scale-card to="https://example.com" target="_blank" rel="noopener noreferrer"  disabled="true" data-sketch-key="card-15">
                <div>
                    <div style="margin: -24px;">
                        <img src="/placeholder.png" alt="placeholder" style="margin-bottom: -4px; width: 100%;">
                        <div style="padding: 24px;">A card with a full width image</div>
                    </div>
                </div>
            </scale-card>
        </div>
    </div>

</div>


<script src="/symbol_names.js"></script>
<script>
    Array.from(document.querySelectorAll("scale-card")).forEach(btn => {
        btn.dataset.sketchSymbol = `Card / 02 Interactive / ${
                order(2, 'Padding ' + (btn.dataset.noPadding == null ? 'on' : 'off'))
            } / ${
                order(4, getStateName(btn))
            }`;
    });
</script>