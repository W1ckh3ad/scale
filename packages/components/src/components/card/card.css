.card {
  width: 100%;
  overflow: hidden;
  box-sizing: border-box;
  background: var(--color-text-contrast, #ffffff);
  transition: var(--transition-generic, all 0.2s ease-in-out);
  border-radius: var(--radius-card, 0.5rem);
  box-shadow: var(
    --shadow-card,
    0 4px 16px 0 rgba(0, 0, 0, 0.1),
    0 2px 4px 0 rgba(0, 0, 0, 0.1)
  );
  /* a11y: add border for high contrast mode */
  border: 1px solid transparent;
}

.card__body {
  padding: var(--spacing-box, 1.5rem);
}

.card__body ::slotted(*) {
  margin: 0;
}

.card--interactive {
  color: inherit;
  cursor: pointer;
  display: block;
  outline: none;
  text-decoration: none;
}

.card--interactive:hover {
  box-shadow: var(
    --shadow-card-hover,
    0 6px 24px 0 rgba(0, 0, 0, 0.1),
    0 3px 6px 0 rgba(0, 0, 0, 0.1)
  );
}

.card--interactive:focus {
  box-shadow: var(
    --shadow-card-focus,
    0 6px 24px 0 rgba(0, 0, 0, 0.1),
    0 3px 6px 0 rgba(0, 0, 0, 0.1),
    0 0 0 2px #009de0;
  );
}

.card--interactive:active {
  border: none;
  box-shadow: var(
    --shadow-card-active,
    0 2px 8px 0 rgba(0, 0, 0, 0.1),
    0 1px 2px 0 rgba(0, 0, 0, 0.1) ;
  );
}

.card--interactive .card__body ::slotted(*) {
  margin: 0;
}
