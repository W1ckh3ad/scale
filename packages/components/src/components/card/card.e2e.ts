import { newE2EPage } from '@stencil/core/testing';

describe('scale-card', () => {
  it('renders', async () => {
    const page = await newE2EPage();
    await page.setContent('<scale-card>Click me!</scale-card>');
    const element = await page.find('scale-card');
    expect(element).toHaveClass('hydrated');
  });
});
