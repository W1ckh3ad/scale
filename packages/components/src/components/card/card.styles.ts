import { JssStyle } from 'jss';
export const styles: JssStyle = {
  card: {
    boxSizing: 'border-box',
    overflow: 'hidden',
    background: ({ color }) => color.white,
    width: '100%',
    transition: ({ transition }) => transition.generic,
    borderRadius: ({ radii }) => radii.card,
    boxShadow: ({ shadow }) => shadow.card,
    border: 'none',
  },
  card__body: {
    padding: ({ spacing }) => spacing.box,
    '& ::slotted(*)': {
      margin: 0,
    },
  },
  'card--no-padding': {
    '& $card__body': {
      padding: 0,
    },
  },
  'card--interactive': {
    color: 'inherit',
    textDecoration: 'none',
    cursor: 'pointer',
    display: 'block',
    outline: 'none',
    '&:hover': {
      boxShadow: ({ shadow }) => shadow.card_hover,
    },
    '&:focus': {
      boxShadow: ({ shadow, size, color }) =>
        `${shadow.card_hover},0 0 0 ${size.border_focus}px ${color.focus}`,
    },
    '&:active': {
      border: 'none',
      boxShadow: ({ shadow }) => shadow.card_active,
    },
    '& $card__body ::slotted(*)': {
      margin: 0,
    },
  },
  'card--disabled': {
    '&, &:hover, &:focus, &:active': {
      opacity: ({ opacity }) => opacity.disabled,
      cursor: 'not-allowed',
      outline: 'none',
    },
  },
};
