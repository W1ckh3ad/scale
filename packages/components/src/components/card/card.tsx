import { Component, Prop, h, Host } from '@stencil/core';
import classNames from 'classnames';
@Component({
  tag: 'scale-card',
  styleUrl: 'card.css',
  shadow: true,
})
export class Card {
  /** (optional) Link card */
  @Prop() to?: string = '';
  /** (optional) Label of the card */
  @Prop() label?: string = '';
  /** (optional) Link card target */
  @Prop() target?: string = '_self';
  /** (optional) Link card rel */
  @Prop() rel?: string = '';
  /** (optional) Injected CSS styles */
  @Prop() styles?: string;

  @Prop() clickCard?: any;

  render() {
    const Tag = !!this.to ? 'a' : 'div';

    return (
      <Host>
        {this.styles && <style>{this.styles}</style>}
        <div class="card-border">
          <Tag
            class={this.getCssClassMap()}
            onClick={this.clickCard}
            {...(!this.to ? { role: 'group' } : {})}
            {...(!!this.to ? { href: this.to } : {})}
            {...(!!this.target ? { target: this.target } : {})}
            {...(!!this.rel ? { rel: this.rel } : {})}
            {...(!!this.label ? { ['aria-label']: this.label } : {})}
          >
            <div class="card__body">
              <slot />
            </div>
          </Tag>
        </div>
      </Host>
    );
  }

  getCssClassMap() {
    return classNames('card', !!this.to && 'card--interactive');
  }
}
