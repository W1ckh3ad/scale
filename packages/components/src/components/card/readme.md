# scale-card

<!-- Auto Generated Below -->


## Usage

### Card

# Default

<scale-card>
  <h1>Card Title</h1>
  <p>This is card content</p>
</scale-card>


### Card-interactive

### Interactive

<scale-card href="http://example.com" target="_blank">
  <p>click me</p>
</scale-card>



## Properties

| Property      | Attribute      | Description                    | Type      | Default     |
| ------------- | -------------- | ------------------------------ | --------- | ----------- |
| `customClass` | `custom-class` | (optional) Card class          | `string`  | `''`        |
| `disabled`    | `disabled`     | (optional) Link disabled       | `boolean` | `false`     |
| `href`        | `href`         | (optional) Link card           | `string`  | `''`        |
| `interactive` | `interactive`  | (optional) Link interactive    | `boolean` | `false`     |
| `label`       | `label`        | (optional) Label of the card   | `string`  | `''`        |
| `noPadding`   | `no-padding`   | (optional) Padding off         | `boolean` | `false`     |
| `styles`      | `styles`       | (optional) Injected jss styles | `any`     | `undefined` |
| `target`      | `target`       | (optional) Link card target    | `string`  | `'_self'`   |


## Dependencies

### Used by

 - [scale-chart-stack-card](../chart-stack-card)

### Graph
```mermaid
graph TD;
  scale-chart-stack-card --> scale-card
  style scale-card fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
