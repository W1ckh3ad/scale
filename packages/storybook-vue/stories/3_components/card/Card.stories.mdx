import { Meta, Story, Props, Preview, Description } from '@storybook/addon-docs/blocks';
import ScaleCard from './ScaleCard.vue';

<!--
  ðŸ‘€

  We're missing two key things from Stencil's docs in the generated readme file:
  - whether attributes are optional or required
  - any description for those attributes
  - plus, writing them again in the .vue file can be error-prone.

  Other minor pain-points:
  - writing the code-snippet manually, need to make sure attribute/values match
  - declaring all props both in Template and the <template> of the .vue file

  Two more thing to think about:
  - in the end there's nothing Vue-specific in the stories, only for us devs, no?
    al least until the code snippet in the Preview works as expected and doesn't show `Template.bind()`
  - we're not using the Vue output target package (it's not really useful in Storybook)
-->

<Meta
  title='Components/Card'
  component={ScaleCard}
  argTypes={{
    label: {
      control: { type: 'text' }
    },
    styles: {
      control: { disable: true }
    }
  }}
/>

export const Template = (args, { argTypes }) => ({
  components: { ScaleCard },
  props: {
    slot: String,
    ...ScaleCard.props
  },
  template: `
    <scale-card
      :custom-class="customClass"
      :href="href"
      :target="target"
      :interactive="interactive"
      :no-padding="noPadding"
      :disabled="disabled"
      :styles="JSON.stringify(styles)"
    >
      {{ slot }}
    </scale-card>
  `
})

# Card

## Standard

<Preview withSource="none">
  <Story
    name='Standard'
    args={{
        slot: 'A card',
        label: 'Example Card',
    }}
  >
    {Template.bind({})}
  </Story>
</Preview>

<Props story='Standard' />

## With Link

<Preview withSource="none">
  <Story
    name='With Link'
    args={{
      slot: 'A card with a link to example.com',
      label: 'Example Card',
      href: 'https://example.com',
      target: '_blank'
    }}
  >
    {Template.bind({})}
  </Story>
</Preview>

```html
<scale-card href="https://example.com" target="_blank" label="Example Card">A card with a link to example.com</scale-card>
```


## Padding Off

<Preview withSource="none">
  <Story
    name='Padding Off'
    args={{
      slot: 'A card with a link to example.com',
      label: 'Example Card',
      noPadding: true
    }}
  >
    {Template.bind({})}
  </Story>
</Preview>

```html
<scale-card href="https://example.com" target="_blank" no-padding label="Example Card">A card with no padding</scale-card>
```

## Interactive

<Preview withSource="none">
  <Story
    name='Interactive'
    args={{
      slot: 'A card, interactive, no link',
      label: 'Example Card',
      interactive: true
    }}
  >
    {Template.bind({})}
  </Story>
</Preview>

```html
<scale-card interactive label="Example Card">A card, interactive</scale-card>
```


## Interactive Padding Off

<Preview withSource="none">
  <Story
    name='Interactive Padding Off'
    args={{
      slot: 'A card, interactive, no link',
      label: 'Example Card',
      interactive: true,
      noPadding: true
    }}
  >
    {Template.bind({})}
  </Story>
</Preview>

```html
<scale-card interactive no-padding label="Example Card">A card, interactive, no link</scale-card>
```

## Disabled

<Preview withSource="none">
  <Story
    name='Disabled'
    args={{
      slot: 'A card, disabled',
      label: 'Example Card',
      disabled: true
    }}
  >
    {Template.bind({})}
  </Story>
</Preview>

```html
<scale-card disabled label="Example Card">A card, interactive</scale-card>
```
