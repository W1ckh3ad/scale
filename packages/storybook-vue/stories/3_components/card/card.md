# Card

Cards group related pieces of information together and represent one conceptional unit. They make it easier for users to quickly scan content and find relevant information.

## General

Cards can hold any kind of content in a flexibly sized container. One card communicates one message. A card is an independent piece of content that requires no additional context.

![Image Name](assets/3_components/card/cards_general.png)

### When to use

Cards are best suited to:

*	encourage and enable users to browse your content (e.g., shop, social media, etc.) or
*	provide summaries of different kinds of content (e.g., dashboards).

### When not to use

Don't use cards:

*	solely for aesthetic reasons or
*	as purely navigational elements.


## Basic types

### Card with an interactive container

The card acts as a preview and point of entry for more detailed information. The card itself is the primary interaction area. Additional interaction elements are not needed. Including additional interactive elements, like buttons or text links, will generate a second action area.

![Image Name](assets/3_components/card/card-interactive.png)

### Card with a static container

A card with a static container also communicates one important message. Generally, no crucial information follows after the card. Although the container is not an interactive component, it can still contain interactive elements like functions or text links to related information.

![Image Name](assets/3_components/card/cards-static.png)

## Layout

### Card Padding

By default, cards have an inner padding to align typography and visual elements within the card. If you want to place images in the bleed, turn the padding off.

![Image Name](assets/3_components/card/cards-padding.png)

## Best practice

### Decide on one type of card

Decide whether you want to use cards with an interactive or static container in your product or service. Never mix card types, because doing so is confusing for users.

### Be consistent

Display content of the same type (e.g., blogposts, products, or cards) with a uniform structure. This provides a balanced appearance so users can easily compare content.

### Design your own card

You can design cards for a variety of situations. The purpose of the card component is to give you the flexibility to build your card according to our principles â€“ as best suits your particular use case.

## Related components

Divider, Data Table
