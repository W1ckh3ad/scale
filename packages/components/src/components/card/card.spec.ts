/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { newSpecPage } from '@stencil/core/testing';
import { Card } from './card';

describe('Card', () => {
  it('should match snapshot', async () => {
    const page = await newSpecPage({
      components: [Card],
      html: `<scale-card>Label</scale-card>`,
    });
    expect(page.root).toMatchSnapshot();
  });
});
