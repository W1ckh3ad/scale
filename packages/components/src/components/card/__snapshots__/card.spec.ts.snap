// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should match snapshot 1`] = `
<scale-card>
  <mock:shadow-root>
    <style>
      .card__border {
  border: 1px solid transparent;
}
.card {
  width: 100%;
  overflow: hidden;
  box-sizing: border-box;
  background: #FFFFFF;
  transition: all .2s ease-in-out;
  border-radius: 0.5rem;
  box-shadow: 0 4px 16px 0 rgba(0,0,0,0.10), 0 2px 4px 0 rgba(0,0,0,0.10);
}
.card__body {
  padding: 1.5rem;
}
.card__body ::slotted(*) {
  margin: 0;
}
.card--interactive {
  color: inherit;
  cursor: pointer;
  display: block;
  outline: none;
  text-decoration: none;
}
.card--interactive:hover {
  box-shadow: 0 6px 24px 0 rgba(0,0,0,0.10), 0 3px 6px 0 rgba(0,0,0,0.10);
}
.card--interactive:focus {
  box-shadow: 0 6px 24px 0 rgba(0,0,0,0.10), 0 3px 6px 0 rgba(0,0,0,0.10),0 0 0 2px #009DE0;
}
.card--interactive:active {
  border: none;
  box-shadow: 0 2px 8px 0 rgba(0,0,0,0.10), 0 1px 2px 0 rgba(0,0,0,0.10);
}
.card--interactive .card__body ::slotted(*) {
  margin: 0;
}
    </style>
    <div class="card__border">
      <div class="card" role="group" target="_self">
        <div class="card__body">
          <slot></slot>
        </div>
      </div>
    </div>
  </mock:shadow-root>
  Label
</scale-card>
`;
